Fix empty worker tags

aholyoake-bc opened the pull request:
Empty worker tags should be treated the same as not having any tags at all
This makes having the CONCOURSE_TAGS being unset equivalent to CONCOURSE_TAGS being set and empty

This change should probably be handled in the environment variable wiring, but it's getting late and there's a lot of magic going on in there

## What does this PR accomplish?

Bug Fix

closes #6056 

## Changes proposed by this PR:
If the tags list has length 1, but the only item is the empty string, then don't fail the match

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

chenbh commented:
Looks like there was a previous attempted fix at https://github.com/concourse/concourse/pull/5355 and https://github.com/concourse/concourse/issues/5354

aholyoake-bc commented:
#5355 looks like a sensible solution - any idea why it was closed?
#5354 solves a related problem of placing jobs on workers with empty tags, rather than configuring the workers themselves.

The question here is: is there any reason to consider an empty worker tag different to not having a worker tag at all?

muntac commented on atc/worker/worker.go:
```diff
@@ -850,7 +850,7 @@ func (worker *gardenWorker) Uptime() time.Duration {
 func (worker *gardenWorker) tagsMatch(tags []string) bool {
 	workerTags := worker.dbWorker.Tags()
-	if len(workerTags) > 0 && len(tags) == 0 {
+	if len(workerTags) > 0 && !(len(workerTags) == 1 && workerTags[0] == "") && len(tags) == 0 {
```
```suggestion
if len(tags) == 0 {
		// If worker only has an empty tag due to user specifying "" instead of []
		if len(workerTags) == 1 && workerTags[0] == "" {
			return true
		}
		if len(workerTags) > 0 {
			return false
		}
	}
```

muntac reviewed (changes requested):
@aholyoake-bc To answer your question - probably no. I tried hard to think of a way to do this at the environment variable level. I could only think of making a custom flag for the worker tags that would treat empty string as null, and then add it to the https://github.com/concourse/flag repo. But it seems like more trouble than it's worth.

We can go with this route. I made a small suggestion so that the condition is easier to read.

I would also add a test case somewhere around this [spot](https://github.com/concourse/concourse/blob/759aa59869586b1b3a974b9f101a52ff3e7973d8/atc/worker/worker_test.go#L566). Let me know if you need help with that.

aholyoake-bc committed:
atc: worker: Fix empty worker tags

Empty worker tags should be treated the same as not having any tags at
all

This makes having the CONCOURSE_TAGS being unset equivalent to
CONCOURSE_TAGS being set and empty

This should probably be handled in the environment variable parsing...

Signed-off-by: Alexander Holyoake <aholyoake@blockchain.com>
